### DISCUSSION TITLE
Envoy troubleshooting - Design discussion

### DISCUSSION THREAD

**Discussion by lucasponce** [kiali/kiali#4082] [Roadmap]
Hi everyone,

Envoy is the cornerstone of the data plane architecture of Istio [1] with very powerful capabilities but, on the other hand, complex to configure and troubleshoot [2].

One of the Kiali’s goal is to help users to understand how the Service Mesh is working and help them to understand from a high level into a low level detail what is happening when something doesn’t seem to work as expected https://github.com/kiali/kiali/issues/4066 .

Kiali has multiple features that help users to understand, debug and troubleshoot Istio Service Mesh:
- Visualize the status of the traffic.
- Validate Istio configuration.
- Access to the workload logs and correlate applications and envoy log together.
- Access to the control plane logs and help to understand envoy log structure.
- Visualize status of synchronization between control plane and data plane proxies as part of the workload health.
- Visualize the envoy clusters, listeners and routes from a workload and link them with Istio traffic definitions.
- Fetch status metrics of Envoy sidecar at workload level.

This area is in quite active development https://github.com/kiali/kiali-ui/pull/2170 and one of the goals is to improve this information, connect them in a way that is easy to understand for a non specialist Envoy user and correlate with rest of telemetry collected to offer a global view about what is happening.

This issue is to collect thoughts and ideas from users that may help Kiali to improve the Envoy troubleshooting scenarios.

Please, join us in these discussions :)

[1] https://istio.io/latest/docs/ops/deployment/architecture/#envoy
[2] https://www.envoyproxy.io/docs/envoy/v1.17.1/intro/arch_overview/arch_overview

**Comment by xeviknal** [comment]
I had this in my head for a little while but today I came across this again. I think it is worth it to explore.
Envoy has the optional admin interface that allows you to view configuration and statistics, change the behavior of the server, and tap traffic according to specific filter rules. By default, envoy proxies deployed by istio are enabled (that is why we have access to the `/config_dump` endpoint).

Please check out what is possible to do through the admin interface[1][2]

Here's a picture of the console with all the operations to perform:
![Screenshot of Envoy Admin](https://user-images.githubusercontent.com/613814/121884758-97d01000-cd13-11eb-96b9-e9095c844bfe.png)

My proposal would be to investigate all the available options in that interface and see if we can either digest some metrics, make it easier available (without the user having to locally forward request), show config info (e.g. server_info) or something really istio-envoy specific.  

[1] https://www.envoyproxy.io/docs/envoy/latest/start/quick-start/admin (QUICK START)
[2] https://www.envoyproxy.io/docs/envoy/latest/operations/admin (DOCS)

**Reply by lucasponce** [reply]
But note that the Istio Proxy is a limited/modified version from Envoy.
Instead to offer a "generic" Envoy admin, I'd suggest to propose a more "user driven" scenarios (i.e. How to debug policy enforcement in Envoy, how to enrich visibility for filtered traffic, etc).

We won't be able to cover all use cases, but I'd like to start for the most value ones.

I've already investigated some of them (specially related for "health" endpoints, but also I've found that it requires a sync with the Proxy version from Istio, as for performance reasons not everything is available).

But +1, with the comment that make it "user driven" instead to map existing APIs.

**Comment by leandroberetta** [comment]
Seeing the debugging/troubleshooting information related to Envoy [1] and the information Kiali already has (clusters, listeners, routes, etc) it would be useful to to the same "walkthrough" that Istio recommends but in a more graphical way (with a relations graph between Envoy entities), for example following the same order that Istio's site suggest, step by step, but in Kiali. 

[1] [https://istio.io/latest/docs/ops/diagnostic-tools/proxy-cmd/](https://istio.io/latest/docs/ops/diagnostic-tools/proxy-cmd/)

**Comment by leandroberetta** [comment]
An interesting thing to see is more information at a glance for example:

- How many retries needed a request to success or fail (with that insight someone can increase the retries or fix a problematic service if that is a possibility, if it is an external service increasing the retries will be the only way to interact with it)
- Same for timeouts, is the timeout set it right?
- More information about circuit breakers, open/close status with a graphical dashboard

**Reply by abonas** [reply]
do you mean to check/display those things "on average" for each microservice (which is typically group of pods/workload)? per single pod for its each respective envoy? compared as an overview for the entire mesh?

**Reply by leandroberetta** [reply]
At workload level, per single pod for its envoy, but would be useful to have this information at service level, aggregating or mixing all related envoys of that service.

Example with circuit breakers:

Workload level: the circuit breaker is open, closing in x time
Service level: one of three (random number) circuit breakers are open

**Comment by nrfox** [comment]
Related to exposing more functionality from the envoy admin api: https://github.com/kiali/kiali/discussions/4082#discussioncomment-867612

Having a drop down on the workload --> logs tab where you can select the log level for the envoy proxy would be useful when debugging a particular workload. It's possible to select the log level of individual loggers through the envoy admin api but even just being able to set the global log level in Kiali would be useful.

**Reply by abonas** [reply]
Do you mean setting global log level for all envoys in the mesh? because you mentioned drop down on a workload and that means specific envoy's log level, so it was not clear what is the suggestion

**Reply by nrfox** [reply]
No just for that specific workload. The goal being if you a trying to debug a particular workload/service you can easily get more logs from the proxy by selecting the log level drop down. At debug level the proxy logs are much more helpful in determining what is happening to a request but you probably only want to enable this for the particular workload(s) that you are debugging and not the entire mesh.

**Comment by jmazzitelli** [comment]
Another feature idea - when a check of a proxy-status determines something is not synced, show a diff of the changes that are not synced. See: https://istio.io/latest/docs/ops/diagnostic-tools/proxy-cmd/#retrieve-diffs-between-envoy-and-istiod

**Reply by lucasponce** [reply]
And highlight it on the yaml editor of the Envoy could be a plus

**Comment by jmazzitelli** [comment]
I'm reading the Envoy docs to get a better feel for how it is configured. One thing that I can see that would be really helpful is to graphically represent the envoy config (the listeners and routes and maybe clusters) in a similar way we display the workload graph (we could even see if we could decorate our existing workload graph with this information - or overlay an envoy graph over top of it?).

Seeing a graphical visualization of the envoy topology would be extremely helpful.

I can then conceive of features such as clicking a node or edge and obtaining the envoy config related to that portion of the graph (so you could see, for example, the filters involved).

**Reply by lucasponce** [reply]
The idea to combine information graphically sounds good, but two considerations: 
- It's on demand per Pod.
- The info can be vast, so I think it could be good to have it more problem-solving driven, just to put an example, I saw cases like [1] that perhaps Kiali can help to identify and highlight.

[1] https://github.com/istio/istio/issues/32673

**Reply by lucasponce** [reply]
Another idea, in most of the cases, when Envoy has a problem, it typically means no traffic, so that's also to take it into consideration on correlation use cases.
